### PR TITLE
Close #268: Support Github like task list

### LINKED ISSUE #268: Support Github-like task list
Reference: https://help.github.com/en/articles/about-task-lists

```
1. [ ] Item 1
1. [x] Item 2
    * [ ] Item 3
    * [ ] Item 4
```

### PR DESCRIPTION
**Pull request by tomyeh**
Introduced a parameter called `checkable` to `Document`. It is false by default. That is, the checkbox will be disabled. If `true` is specified, the checkbox is enabled and an additional attribute called `data-line` is generated with the line number in the source markdown.

### CODE REVIEW

File: lib/src/block_parser.dart [hunk 1]
```diff
@@ -713,9 +721,36 @@ abstract class ListSyntax extends BlockSyntax {
...
+      }
+
+      var itemParser = BlockParser(item.lines, parser.document, item.offset);
       var children = itemParser.parseLines();
-      itemNodes.add(Element('li', children));
+      if (hasCheck) {
+        final check = Element('input', null);
+        if (checked) check.attributes['checked'] = 'checked';
+        if (parser.document.checkable)
+          check.attributes['data-line'] = '${item.offset}';
```
Comment (srawlins) [inline lib/src/block_parser.dart]: I don't see this `data-line` feature on GitHub. I don't think we should add features or complexity beyond what a popular client (like GitHub or CommonMark) has.

File: lib/src/document.dart [hunk 2]
```diff
@@ -13,7 +13,7 @@ class Document {
   final ExtensionSet extensionSet;
   final Resolver linkResolver;
   final Resolver imageLinkResolver;
-  final bool encodeHtml;
+  final bool encodeHtml, checkable;
```
[the remark this answers is not in the document]
Comment (srawlins) [reply lib/src/document.dart]: Also `checkable` definitely deserves its own doc comment.

Comment (srawlins) [inline lib/src/document.dart]: The checkboxes feature should only be enabled via a separate Syntax, added by default in the `gitHubWeb` ExtensionSet.

File: lib/src/block_parser.dart (hunk 1 above)
Comment (tomyeh) [reply lib/src/block_parser.dart]: When using a checklist, the app needs to modify the markdown when user clicks on the checkbox. By generating the line number, it is easier for app to write such kind of code. Or, we need another complicated utility to locate `[ ]` from the markdown.

File: lib/src/block_parser.dart [hunk 3]
```diff
@@ -738,12 +773,12 @@ abstract class ListSyntax extends BlockSyntax {
       }
     }
 
-    if (listTag == 'ol' && startNumber != 1) {
-      return Element(listTag, itemNodes)..attributes['start'] = '$startNumber';
-    } else {
-      return Element(listTag, itemNodes);
-    }
+    final ul = Element(listTag, itemNodes);
```
[the remark this answers is not in the document]
Comment (tomyeh) [reply lib/src/block_parser.dart]: It is a simple optimization. No functional difference.

[the remark this answers is not in the document]
Comment (KagurazakaHanabi) [reply lib/src/block_parser.dart]: Maybe should named `element` or `el`?

File: lib/src/block_parser.dart [hunk 4]
```diff
@@ -738,12 +773,12 @@ abstract class ListSyntax extends BlockSyntax {
...
+    if (listTag == 'ol' && startNumber != 1)
+      ul.attributes['start'] = '$startNumber';
+    return ul;
   }
+  static final _reCheckbox = RegExp(r'^\[([x ])\](\s+\S.+)$');
```
[the remark this answers is not in the document]
Comment (KagurazakaHanabi) [reply lib/src/block_parser.dart]: `[x ]` -> `[Xx ]`

**Pull request closed without merging**